add docs for mecanum rovers (#3411)
PX4/PX4-user_guide: en/config_rover/mecanum.md

# Configuration/Tuning (Mecanum Rover)

This topic provides a step-by-step guide for setting up your [Mecanum rover](../frames_rover/mecanum.md).
Successive steps enable [drive modes](../flight_modes_rover/mecanum.md) with more autopilot support and features.

::: warning
Each step is dependent on the previous steps having been completed.
Modes will only work properly if the preceding modes have been configured.
:::

## Basic Setup

To start using the mecanum rover:

1. Enable Rover support by flashing the [PX4 rover build](../frames_rover/index.md#flashing-the-rover-build) onto your flight controller.
   Note that this is a special build that contains rover-specific modules.

2. In the [Airframe](../config/airframe.md) configuration select _Generic Rover Mecanum_ frame:

   ![QGC screenshot showing selection of the airframe 'Generic Rover Mecanum'](../../assets/config/airframe/airframe_generic_rover_mecanum.png)

   Select the **Apply and Restart** button.

   ::: info
   If this airframe does not show up in the UI, it can alternatively be selected by setting the [SYS_AUTOSTART](../advanced_config/parameter_reference.md#SYS_AUTOSTART) parameter to `52000`.
   :::

> 3. Use [Actuators Configuration & Testing](../config/actuators.md) to map the motor functions to flight controller outputs.

Reviewer: Is there anything special about this? I mean is it worth having a screenshot or two of a standard setup, or it is trivially easy for anyone to set up reading existing docs?

Reviewer: Great. Sometimes it is worth showing an image for this kind of thing to show how trivial it is. Also helps answer questions like "what sort of ESCs should I use". For an MC that is ideally D-Shot, whereas for an FW it is often PWM.

## Manual Mode

The basic setup (above) is all that is required to use the rover in [Manual mode](../flight_modes_rover/mecanum.md#manual-mode).

::: info
In manual mode the stick inputs are directly mapped to motor commands.
Especially moving the stick that controls the yaw rate all the way to one side will cause the wheels on the left and right to spin at full speed in opposite directions.
Depending on the rover this can lead to a very aggressive rotation.
The parameter [RM_MAN_YAW_SCALE](#RM_MAN_YAW_SCALE) can be used to scale the manual input for the yaw rate.
By reducing the parameter from the default value of 1 this behaviour can be tuned.
Note that this parameter only affects this mode, not any of the following ones.
:::

## Acro Mode

To set up [Acro mode](../flight_modes_rover/mecanum.md#acro-mode) navigate to [Parameters](../advanced_config/parameters.md) in QGroundControl and set the following parameters:

> 1. [RM_WHEEL_TRACK](#RM_WHEEL_TRACK) [m]: Measure the distance from the centre of the right wheel to the centre of the left wheel.

Reviewer: I presume this can't be inferred from the vehicle geometry in the actuators screen?

Author: This would be a nice feature, but it is currently not implemented. I'll look into adding it.

   ![Wheel track](../../assets/airframes/rover/rover_differential/wheel_track.png)

> 2. [RM_MAX_YAW_RATE](#RM_MAX_YAW_RATE) [deg/s]: This is the maximum yaw rate you want to allow for your rover.

Reviewer: Is there advice for a good initial value? What happens if you get it wrong?

Author: Elaborated on it in https://github.com/PX4/PX4-user_guide/pull/3411/commits/6474c87f6d8345975b93dafd472d8e9d18fbf632

Reviewer: Nice! Subedited (and ran prettier) in https://github.com/PX4/PX4-user_guide/pull/3411/commits/07a21f52161798ec4aba89dc72987d911959b0c6

   This will define the stick-to-yaw-rate mapping for all manual modes using closed loop yaw control and set an upper limit for the yaw rate setpoint for all [auto modes](#auto-modes).
3. [RM_MAX_THR_YAW_R](#RM_MAX_YAW_RATE) [m/s]: This parameter is used to calculate the feed-forward term of the closed loop yaw rate control.
   The controller calculates the required speed difference between the left and right motor to achieve the desired yaw rate.

>    This desired speed difference is then linearly mapped to normalized motor commands.
>    To get a good starting value for this parameter drive the rover in manual mode forwards at full throttle and note the ground speed of the vehicle.

Reviewer: ```suggestion
   This desired speed difference is then linearly mapped to normalized motor commands.

   To get a good starting value for this parameter drive the rover in manual mode forwards at full throttle and note the ground speed of the vehicle.
```

   This way the yaw rate is only controlled by the feed-forward term, which makes it easier to tune.
   Now put the rover in [Acro mode](../flight_modes_rover/mecanum.md#acro-mode) and then move the right-stick of your controller to the right and/or left and hold it at a few different levels for a couple of seconds each.
   Disarm the rover and from the flight log plot the _yaw_rate_setpoint_ and actual _yaw_rate_ from the [RoverMecanumSetpoint](../msg_docs/RoverMecanumStatus.md) over each other.
   If the actual yaw rate of the rover is higher than the yaw rate setpoint, increase [RM_MAX_THR_YAW_R](#RM_MAX_YAW_RATE).
   If it is the other way around decrease the parameter and repeat until you are satisfied with the setpoint tracking.
   :::

4. [RM_YAW_RATE_P](#RM_YAW_RATE_P) [-]: Proportional gain of the closed loop yaw rate controller.
   Unlike the feed-forward part of the controller, the closed loop yaw rate control will compare the yaw rate setpoint with the measured yaw rate and adapt to motor commands based on the error between them.
   The proportional gain is multiplied with this error and that value is added to the motor command.

>    This way disturbances like uneven grounds or external forces can be compensated.

Reviewer: ```suggestion
   This compensates for disturbances such as uneven ground and external forces.
```

   ::: tip
   This parameter can be tuned the same way as [RM_MAX_THR_YAW_R](#RM_YAW_RATE_P_TUNING).

>    If you tuned [RM_MAX_THR_YAW_R](#RM_MAX_YAW_RATE) well, you might only need a very small value.

Reviewer: I don't get " This parameter can be tuned the same way as [RM_MAX_THR_YAW_R]"

That says:

> To get a good starting value for this parameter drive the rover in manual mode forwards at full throttle and note the ground speed of the vehicle. Then enter twice this value for the parameter.

So the value is 2 times ground speed. 
So why would that be small and why affected by tuning for `RM_MAX_THR_YAW_R`?

Author: Updated this in https://github.com/PX4/PX4-user_guide/pull/3411/commits/ab30f0666db49e0a694bfe92f2f89ca2f9760945

   :::

5. (Optional) [RM_YAW_RATE_I](#RM_YAW_RATE_I) [-]: Integral gain of the closed loop yaw controller.
   The integral gain accumulates the error between the desired and actual yaw rate over time and that value is added to the motor command.

   ::: tip
   The integrator gain is usually not necessary for the yaw rate setpoint as this is usually a fast changing value.
   Leave this parameter at zero unless necessary, as it can have negative side effects such as overshooting or oscillating around the setpoint.
   :::

The rover is now ready to drive in [Acro mode](../flight_modes_rover/mecanum.md#acro-mode).

## Stabilized Mode

::: warning
For this mode to work properly [Acro mode](#acro-mode) must've already been configured!
:::

> For [Stabilized mode](../flight_modes_rover/mecanum.md#stabilized-mode) the controller utilizes a closed loop yaw controller, which creates a yaw rate setpoint to control the yaw when it is active:

Reviewer: The fact that is is active is implied

```suggestion
For [Stabilized mode](../flight_modes_rover/mecanum.md#stabilized-mode) the controller utilizes a closed loop yaw controller, which creates a yaw rate setpoint to control the yaw.
```

> For [Stabilized mode](../flight_modes_rover/mecanum.md#stabilized-mode) the controller utilizes a closed loop yaw controller, which creates a yaw rate setpoint to control the yaw when it is active:

Reviewer: Maybe
```suggestion
For [Stabilized mode](../flight_modes_rover/mecanum.md#stabilized-mode) the controller utilizes a closed loop yaw controller, which creates a yaw rate setpoint to control the yaw when driving in a straight line (no yaw rate input).
```

Therefore you only need to tune the closed loop gains:

1. [RM_YAW_P](#RM_YAW_P) [-]: Proportional gain for the closed loop yaw controller.

   ::: tip
   In stabilized mode the closed loop yaw control is only active when driving a straight line (no yaw rate input).
   To tune it set [RM_YAW_I](#RM_YAW_I) to zero and start with a value of 1 for [RM_YAW_P](#RM_YAW_P).
   Put the rover into stabilized mode and move the left stick of your controller up and/or down to drive forwards/backwards.
   Disarm the rover and from the flight log plot the _yaw_setpoint_ from the [RoverMecanumSetpoint](../msg_docs/RoverMecanumSetpoint.md) message and the _actual_yaw_ from the [RoverMecanumStatus](../msg_docs/RoverMecanumStatus.md) message over each other.
   Increase/Decrease the parameter until you are satisfied with the setpoint tracking.
   :::

2. [RM_YAW_I](#RM_YAW_I) [-]: Integral gain for the closed loop yaw controller.

   ::: tip
   For the closed loop yaw control an integrator gain is useful because this setpoint is often constant for a while and an integrator eliminates steady state errors that can cause the rover to never reach the setpoint.

>    In [Auto Modes](#auto-modes) there will be a further elaboration on why an integrator is necessary for the yaw controller.

Reviewer: How do we tune this one - you explain the other

Author: Updated this in https://github.com/PX4/PX4-user_guide/pull/3411/commits/ab30f0666db49e0a694bfe92f2f89ca2f9760945

   :::

The rover is now ready to drive in [Stabilized mode](../flight_modes_rover/mecanum.md#stabilized-mode).

## Position Mode

:::warning
For this mode to work properly [Acro mode](#acro-mode) and [Stabilized mode](#stabilized-mode) must already be configured!
:::

[Position mode](../flight_modes_rover/mecanum.md#position-mode) is the most advanced manual mode, utilizing closed loop yaw rate, yaw and speed control and leveraging position estimates.

To configure set the following parameters:

1. [RM_MAX_SPEED](#RM_MAX_SPEED) [m/s]: This is the maximum speed you want to allow for your rover.
   This will define the stick-to-speed mapping for position mode and set an upper limit for the speed setpoint for all [auto modes](#auto-modes).
2. [RM_MAX_THR_SPD](#RM_MAX_SPEED) [m/s]: This parameter is used to calculate the feed-forward term of the closed loop speed control which linearly maps desired speeds to normalized motor commands.
   A good starting point is the observed ground speed when the rover drives at maximum throttle in [Manual mode](../flight_modes_rover/mecanum.md#manual-mode).

>    Increase this parameter if the rover is faster than the setpoint, and decrease if the rover is slower.

Reviewer: I don't really get what this means - if I measure speed at max throttle, how then do I know what the setpoint is to compare?

I wonder if some of this could be auto-tuned in some way?

Author: Updated this in https://github.com/PX4/PX4-user_guide/pull/3411/commits/e9016d6c9799d585384ce3ee6d8bf3a6d76f8254.
Autotuning might be possible, but that is not implemented yet. I'll look into it.

Reviewer: Nice. I subedited in https://github.com/PX4/PX4-user_guide/pull/3411/commits/6bd775bd89645a3f42cddeceb98b5ad240110286 

Mostly this indicates my preference for instructions to use lists.

   If the rover decelerates too quickly decrease the [RM_MAX_ACCEL](#RM_MAX_ACCEL) parameter, if it starts slowing down too early increase the parameter.
   If you observe a jerking motion as the rover slows down, decrease the [RM_MAX_JERK](#RM_MAX_JERK) parameter otherwise increase it as much as possible as it can interfere with the tuning of [RM_MAX_ACCEL](#RM_MAX_ACCEL).
   These two parameters have to be tuned as a pair, repeat until you are satisfied with the behaviour.
   :::

4. [RM_MISS_VEL_GAIN](#RM_MISS_VEL_GAIN): The rover slows down when approaching a waypoint based on the angle between the line segment from the previous to current waypoint and current to next waypoint. How much the rover slows down can be tuned with this parameter:
   
   $$
    v_{transition} = v_{max} \cdot (1 - \theta_{normalized} * k)
   $$
with $v_{transition}:$ Transition velocity, $v_{max}:$  Maximum speed ([RM_MISS_SPD_DEF](#RM_MISS_SPD_DEF)), $\theta_{normalized}:$ Angle between the linesegment of prev-curr and curr-next waypoints normalized from $[0\degree, 180\degree]$ to $[0, 1]$ and $k:$ Tuning parameter ([RM_MISS_VEL_GAIN](#RM_MISS_VEL_GAIN)).

> 

Reviewer: This renders hard to read. Perhaps put the values in a bullet list?

Author: Updated in https://github.com/PX4/PX4-user_guide/pull/3411/commits/c5045f58c7e90e9b80aa7d2641ca1c564476e6bc
